<a href="https://colab.research.google.com/github/Salah0Uddin/Machine-Learning-Full/blob/main/ML_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Decision Tree Regressor and Cross Validation

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:

import urllib.request

url = "http://lib.stat.cmu.edu/datasets/boston"
raw_txt = urllib.request.urlopen(url).read().decode("utf-8").splitlines()

# Skip header lines and parse the data manually
data = []
for i in range(22, len(raw_txt), 2):
    row1 = list(map(float, raw_txt[i].strip().split()))
    row2 = list(map(float, raw_txt[i+1].strip().split()))
    data.append(row1 + row2)

# Create DataFrame
columns = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE',
           'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
df = pd.DataFrame(data, columns=columns)
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  MEDV  
0     15.3  396.90   4.98  24.0  
1     17.8  396.90   9.14  21.6  
2     17.8  392.83   4.03  34.7  
3     18.7  394.63   2.94  33.4  
4     18.7  396.90   5.33  36.2

In [3]:
X=df.iloc[:,:-1]
y=df["MEDV"]


In [4]:
X.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  
0     15.3  396.90   4.98  
1     17.8  396.90   9.14  
2     17.8  392.83   4.03  
3     18.7  394.63   2.94  
4     18.7  396.90   5.33

In [5]:
y.head()

0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
Name: MEDV, dtype: float64

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.3,random_state=42)

In [7]:
regressor=DecisionTreeRegressor()
regressor.fit(X_train,y_train)

DecisionTreeRegressor()

In [8]:
y_pred=regressor.predict(X_test)
y_pred

array([21.6, 33.1, 13.3, 24.8, 19.4, 18.5, 22.5, 14.3, 23. , 21.2, 27.1,
       27.1,  5. , 21.4, 18.7, 25. , 15.4,  7.2, 48.3, 15.2, 24.1, 23.1,
       13.6, 22. , 12.7, 14.6, 22. , 14.1, 21.7, 18.9, 18.8, 23.1, 10.4,
       23.1, 16.7, 15.6, 33.1, 18.7, 20.4, 24.8, 19.8, 32. , 50. , 16.2,
       21.6, 12.7, 13.6, 24.8, 20. , 32.5, 22.9, 33.4, 21.7, 28.4, 43.1,
       19.5, 16.8, 27.9, 21.6, 22.5, 24.5, 33. , 29.4, 18.2, 28. , 13.6,
       15.4, 22.9, 28. , 14.1, 22.6, 28.7, 11.8, 23.7, 21.4,  5.6, 19.3,
       44. , 13.1,  8.1, 22. , 11.9, 19.3, 11.8, 20.3, 30.1, 14.9, 23.1,
       28.7, 18. , 22.6,  5. , 19.8, 18.5, 24.3, 19.2, 50. , 11.9, 16.7,
       14.5, 17.5, 28.1, 16.7, 20.4, 21.2, 10.2, 20.4, 24.8, 17.5, 28.7,
       11.8, 14.5, 22.2, 29.6, 31.7, 13.5, 50. , 14.8, 21.2, 23.7, 17.1,
       24.8,  5. , 19.3, 24.7, 23.1, 28.7, 37.2, 21.4, 43.8, 15.4, 25. ,
       18.2, 18.4, 11.8, 22.2, 17.5, 31.2, 29.8, 16. , 22.5, 23.7, 19.3,
       16.6,  5. , 18.5, 16.8,  8.5, 13.6, 44.8, 13

In [14]:
from sklearn.metrics import r2_score
r2_score(y_test,y_pred)

0.8446608271557241

In [16]:
#hyperparameter tuning
params={'max_depth':[2,3,4,5,6,7,8,9,10,11,12,13,14,15],
        'criterion':['squared_error','absolute_error','poisson','friedman_mse'],
        'splitter':['best','random'],
        'min_samples_split':[2,3,4,5,6,7,8,9,10,11,12,13,14,15],
        'max_features':['auto','sqrt','log2']

        }

regressor2=DecisionTreeRegressor()

In [17]:
from sklearn.model_selection import GridSearchCV
gridcv=GridSearchCV(regressor2,params,cv=15,scoring="neg_mean_squared_error")


In [18]:
gridcv.fit(X_train,y_train)

GridSearchCV(cv=15, estimator=DecisionTreeRegressor(),
             param_grid={'criterion': ['squared_error', 'absolute_error',
                                       'poisson', 'friedman_mse'],
                         'max_depth': [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
                                       14, 15],
                         'max_features': ['auto', 'sqrt', 'log2'],
                         'min_samples_split': [2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
                                               12, 13, 14, 15],
                         'splitter': ['best', 'random']},
             scoring='neg_mean_squared_error')

In [19]:
gridcv.best_params_

{'criterion': 'friedman_mse',
 'max_depth': 12,
 'max_features': 'log2',
 'min_samples_split': 11,
 'splitter': 'best'}

In [21]:
y_pred=gridcv.predict(X_test)
y_pred

array([22.41538462, 31.66666667, 15.6       , 23.08333333, 21.93333333,
       25.12857143, 23.52      , 13.3       , 20.12222222, 19.325     ,
       21.9625    , 21.36666667,  9.43      , 25.12857143, 23.52      ,
       23.84      , 20.26666667,  9.37      , 43.94285714, 14.64      ,
       24.04      , 24.04      , 14.7       , 18.85      , 12.78333333,
       12.78333333, 19.325     , 14.575     , 21.9625    , 23.52      ,
       21.9625    , 23.05      , 23.84      , 12.78333333, 12.78333333,
       17.86      , 33.97142857, 18.9       , 20.26666667, 24.04      ,
       23.52      , 27.3       , 43.94285714, 18.08      , 22.41538462,
       12.78333333, 14.7       , 24.04      , 20.12222222, 22.8       ,
       21.36666667, 30.25      , 18.85      , 23.84      , 46.91      ,
       21.93333333, 12.78333333, 27.3       , 22.41538462, 18.08      ,
       23.84      , 32.69      , 29.1       , 11.7       , 29.1       ,
       21.425     ,  9.37      , 24.04      , 26.14      , 14.57

In [22]:
r2_score(y_test,y_pred)

0.6875928927884964

In [25]:
# from sklearn.metrics import confusion_matrix,classification_report
# print(confusion_matrix(y_test,y_pred))
# print(classification_report(y_test,y_pred))

from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
print("Mean Squared Error:", mean_squared_error(y_test, y_pred))
print("Mean Absolute Error:", mean_absolute_error(y_test, y_pred))
print("R-squared:", r2_score(y_test, y_pred))

Mean Squared Error: 23.278420635731653
Mean Absolute Error: 2.9089016612042933
R-squared: 0.6875928927884964


The model is able to explain about 68.8% of the variance in housing prices.With an MAE of ~2.91, the model is off by about $2,910 (in $1,000s) on average.Overall, the model performs reasonably well and serves as a good baseline for regression using decision trees.